File: src/gearbox_sim/control/speed_controller.py
"""Generic speed controller for any drivetrain type."""

from dataclasses import dataclass
from typing import Callable, Dict, List, Optional

import numpy as np

from ..core.drivetrain import Drivetrain
from .base import DrivetrainController


@dataclass
class TorqueAllocation:
    """Describes how to allocate torque demand among actuators."""

    actuators: List[str]  # Component names
    fractions: List[float]  # Fraction of demand for each
    is_engine: List[bool]  # Whether each is an engine (non-negative torque only)


class SpeedController(DrivetrainController):
    """Generic proportional-integral speed controller.

    Works with any drivetrain by discovering actuators (engines, motors)
    and distributing torque demand among them.
    """

    def __init__(
        self,
        drivetrain: Drivetrain,
        Kp: float = 50000.0,
        Ki: float = 5000.0,
        allocation: Optional[TorqueAllocation] = None,
    ):
        """Initialize the speed controller.

        Args:
            drivetrain: Drivetrain to control
            Kp: Proportional gain [N·m/(m/s)]
            Ki: Integral gain [N·m/(m·s)]
            allocation: Torque allocation strategy (auto-detected if None)
        """
        super().__init__(drivetrain)
        self.Kp = Kp
        self.Ki = Ki
        self._integral = 0.0
        self._last_time = 0.0

        # Discover actuators if not specified
        if allocation is None:
            self._discover_actuators()
        else:
            self.allocation = allocation

        # Find gearbox for gear control
        self._gearbox_name = self._find_gearbox()
        self._current_gear = 0

    def _discover_actuators(self) -> None:
        """Discover engines and motors in the drivetrain."""
        actuators = []
        fractions = []
        is_engine = []

        components = self.drivetrain.topology.components

        # Find all engines and motors
        for name, comp in components.items():
            comp_type = type(comp).__name__
            if "Engine" in comp_type:
                actuators.append(name)
                is_engine.append(True)
            elif "Motor" in comp_type:
                actuators.append(name)
                is_engine.append(False)

        if not actuators:
            raise ValueError("No actuators (engines/motors) found in drivetrain")

        # Simple equal allocation
        n = len(actuators)
        fractions = [1.0 / n] * n

        self.allocation = TorqueAllocation(actuators, fractions, is_engine)

    def _find_gearbox(self) -> Optional[str]:
        """Find gearbox component name."""
        for name, comp in self.drivetrain.topology.components.items():
            if "Gearbox" in type(comp).__name__ and comp.n_gears > 1:
                return name
        return None

    def compute(self, state: Dict[str, float], grade: float) -> Dict[str, float]:
        """Compute torque commands for all actuators.

        Args:
            state: Current state
            grade: Road grade

        Returns:
            Control dict with torque commands for all actuators
        """
        # Get current velocity
        velocity = self.get_velocity(state)

        # Get target velocity
        v_target = self._target_velocity if self._target_velocity is not None else 10.0

        # Speed error
        v_error = v_target - velocity

        # PI control
        T_demand = self.Kp * v_error + self.Ki * self._integral

        # Update integral (with anti-windup)
        if abs(self._integral) < 100.0:
            self._integral += v_error * 0.1  # Assuming ~0.1s between calls

        # Allocate torque among actuators
        controls = self._allocate_torque(state, T_demand)

        # Add gear control if gearbox exists
        if self._gearbox_name:
            gear = self._select_gear(velocity, grade)
            controls[f"gear_{self._gearbox_name}"] = gear
            self._current_gear = gear

        return controls

    def _allocate_torque(self, state: Dict[str, float], T_demand: float) -> Dict[str, float]:
        """Allocate torque demand among actuators.

        Args:
            state: Current state
            T_demand: Total torque demand

        Returns:
            Control dict with torque for each actuator
        """
        controls = {}

        for i, actuator in enumerate(self.allocation.actuators):
            T_alloc = T_demand * self.allocation.fractions[i]

            # Engines can only provide positive torque
            if self.allocation.is_engine[i]:
                T_alloc = max(0.0, T_alloc)

            # Clip to actuator limits
            component = self.drivetrain.get_component(actuator)
            if hasattr(component, "clip_torque"):
                # Get actuator speed
                speed_key = f"{actuator}.shaft"
                omega = 0.0
                for key, value in state.items():
                    if actuator in key:
                        omega = abs(value)
                        break

                rpm = omega * 30.0 / np.pi
                T_alloc = component.clip_torque(rpm, T_alloc)

            controls[f"T_{actuator}"] = T_alloc

        return controls

    def _select_gear(self, velocity: float, grade: float) -> int:
        """Simple gear selection based on velocity.

        Args:
            velocity: Current velocity [m/s]
            grade: Road grade

        Returns:
            Gear index
        """
        if self._gearbox_name is None:
            return 0

        gearbox = self.drivetrain.get_component(self._gearbox_name)
        n_gears = gearbox.n_gears

        # Simple speed-based selection
        # Each gear covers a velocity range
        v_max = 15.0  # Assumed max speed
        gear_width = v_max / n_gears

        gear = int(velocity / gear_width)
        gear = max(0, min(gear, n_gears - 1))

        # Stay in lower gear on steep grades
        if grade > 0.05:
            gear = max(0, gear - 1)
        if grade > 0.10:
            gear = 0

        return gear

    def reset(self) -> None:
        """Reset controller state."""
        self._integral = 0.0
        self._current_gear = 0

File: web/packages/drivetrain-sim/src/control/speed-controller.ts
/**
 * Generic speed controller for any drivetrain type.
 */

import { Drivetrain } from '../core/drivetrain';
import { DrivetrainController } from './base';

/**
 * Describes how to allocate torque demand among actuators.
 */
export interface TorqueAllocation {
  /** Component names */
  actuators: string[];
  /** Fraction of demand for each */
  fractions: number[];
  /** Whether each is an engine (non-negative torque only) */
  isEngine: boolean[];
}

/**
 * Generic proportional-integral speed controller.
 *
 * Works with any drivetrain by discovering actuators (engines, motors)
 * and distributing torque demand among them.
 */
export class SpeedController extends DrivetrainController {
  readonly Kp: number;
  readonly Ki: number;
  allocation!: TorqueAllocation;
  private _integral: number = 0;
  private _gearboxName: string | null = null;
  private _currentGear: number = 0;
  private _vMax: number;

  constructor(
    drivetrain: Drivetrain,
    Kp: number = 50000.0,
    Ki: number = 5000.0,
    allocation?: TorqueAllocation
  ) {
    super(drivetrain);
    this.Kp = Kp;
    this.Ki = Ki;

    // Discover actuators if not specified
    if (allocation) {
      this.allocation = allocation;
    } else {
      this._discoverActuators();
    }

    // Find gearbox for gear control
    this._gearboxName = this._findGearbox();

    // Discover max speed from vehicle component
    this._vMax = this._discoverMaxSpeed();
  }

  private _discoverMaxSpeed(): number {
    for (const comp of this.drivetrain.topology.components.values()) {
      if (comp.componentType === 'vehicle' && 'params' in comp) {
        const vehicleParams = (comp as any).params;
        if ('vMax' in vehicleParams) {
          return vehicleParams.vMax;
        }
      }
    }
    // Default fallback if no vehicle found
    return 15.0;
  }

  private _discoverActuators(): void {
    const actuators: string[] = [];
    const fractions: number[] = [];
    const isEngine: boolean[] = [];

    const components = this.drivetrain.topology.components;

    // Find all engines and motors using componentType (survives minification)
    for (const [name, comp] of components) {
      if (comp.componentType === 'engine') {
        actuators.push(name);
        isEngine.push(true);
      } else if (comp.componentType === 'motor') {
        actuators.push(name);
        isEngine.push(false);
      }
    }

    if (actuators.length === 0) {
      throw new Error('No actuators (engines/motors) found in drivetrain');
    }

    // Simple equal allocation
    const n = actuators.length;
    for (let i = 0; i < n; i++) {
      fractions.push(1.0 / n);
    }

    this.allocation = { actuators, fractions, isEngine };
  }

  private _findGearbox(): string | null {
    for (const [name, comp] of this.drivetrain.topology.components) {
      if (comp.componentType === 'gearbox' && 'nGears' in comp) {
        if ((comp as any).nGears > 1) {
          return name;
        }
      }
    }
    return null;
  }

  compute(state: Record<string, number>, grade: number): Record<string, number> {
    // Get current velocity
    const velocity = this.getVelocity(state);

    // Get target velocity
    const vTarget = this._targetVelocity ?? 10.0;

    // Speed error
    const vError = vTarget - velocity;

    // PI control
    const tDemand = this.Kp * vError + this.Ki * this._integral;

    // Update integral (with anti-windup)
    if (Math.abs(this._integral) < 100.0) {
      this._integral += vError * 0.1; // Assuming ~0.1s between calls
    }

    // Allocate torque among actuators
    const controls = this._allocateTorque(state, tDemand);

    // Add gear control if gearbox exists
    if (this._gearboxName) {
      const gear = this._selectGear(velocity, grade);
      controls[`gear_${this._gearboxName}`] = gear;
      this._currentGear = gear;
    }

    return controls;
  }

  private _allocateTorque(
    state: Record<string, number>,
    tDemand: number
  ): Record<string, number> {
    const controls: Record<string, number> = {};

    for (let i = 0; i < this.allocation.actuators.length; i++) {
      const actuator = this.allocation.actuators[i];
      let tAlloc = tDemand * this.allocation.fractions[i];

      // Engines can only provide positive torque
      if (this.allocation.isEngine[i]) {
        tAlloc = Math.max(0.0, tAlloc);
      }

      // Clip to actuator limits
      const component = this.drivetrain.getComponent(actuator);
      if (component && 'clipTorque' in component) {
        // Get actuator speed using exact state key format
        // Engines and motors have a 'shaft' port
        const shaftKey = `${actuator}.shaft`;
        const omega = Math.abs(state[shaftKey] ?? 0);

        const rpm = (omega * 30.0) / Math.PI;
        tAlloc = (component as any).clipTorque(rpm, tAlloc);
      }

      controls[`T_${actuator}`] = tAlloc;
    }

    return controls;
  }

  private _selectGear(velocity: number, grade: number): number {
    if (this._gearboxName === null) {
      return 0;
    }

    const gearbox = this.drivetrain.getComponent(this._gearboxName) as any;
    const nGears = gearbox.nGears;

    // Simple speed-based selection
    const gearWidth = this._vMax / nGears;

    let gear = Math.floor(velocity / gearWidth);
    gear = Math.max(0, Math.min(gear, nGears - 1));

    // Stay in lower gear on steep grades
    if (grade > 0.05) {
      gear = Math.max(0, gear - 1);
    }
    if (grade > 0.1) {
      gear = 0;
    }

    return gear;
  }

  reset(): void {
    this._integral = 0;
    this._currentGear = 0;
  }
}
